src/target/riscv/riscv.c

```diff
@@ -477,203 +556,227 @@ static int maybe_add_trigger_t1(struct target *target,
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_EQUAL);
-	tdata1 |= CSR_MCONTROL_M;
-	if (r->misa & (1 << ('S' - 'A')))
-		tdata1 |= CSR_MCONTROL_S;
-	if (r->misa & (1 << ('U' - 'A')))
-		tdata1 |= CSR_MCONTROL_U;
-
-	if (trigger->execute)
-		tdata1 |= CSR_MCONTROL_EXECUTE;
-	if (trigger->read)
-		tdata1 |= CSR_MCONTROL_LOAD;
-	if (trigger->write)
-		tdata1 |= CSR_MCONTROL_STORE;
-
-	riscv_set_register(target, GDB_REGNO_TDATA1, tdata1);
-
-	uint64_t tdata1_rb;
-	int result = riscv_get_register(target, &tdata1_rb, GDB_REGNO_TDATA1);
-	if (result != ERROR_OK)
-		return result;
-	LOG_DEBUG("tdata1=0x%" PRIx64, tdata1_rb);
+	RISCV_INFO(r);
 
-	if (tdata1 != tdata1_rb) {
-		LOG_DEBUG("Trigger doesn't support what we need; After writing 0x%"
-				PRIx64 " to tdata1 it contains 0x%" PRIx64,
-				tdata1, tdata1_rb);
-		riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
-	}
+	tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_TYPE(riscv_xlen(target)), 2);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_DMODE(riscv_xlen(target)), 1);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_ACTION, CSR_MCONTROL_ACTION_DEBUG_MODE);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_M, 1);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_S, !!(r->misa & (1 << ('S' - 'A'))));
+	tdata1 = set_field(tdata1, CSR_MCONTROL_U, !!(r->misa & (1 << ('U' - 'A'))));
+	tdata1 = set_field(tdata1, CSR_MCONTROL_EXECUTE, trigger->execute);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_LOAD, trigger->read);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_STORE, trigger->write);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_ANY & 3);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_ANY >> 2) & 3);
+
+	if (trigger->execute || trigger->length == 1)
+		goto MATCH_EQUAL;
+	if (!match_napot_checker(trigger, &tdata2))
+		goto MATCH_GE_LT;
+
+	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
+	if (ret != ERROR_OK)
+		return ret;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK) {
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_GE_LT; /* Fallback to chained MATCH using GT and LE */
+		return ret;
+	}
+	r->trigger_unique_id[idx] = trigger->unique_id;
+	return ERROR_OK;
 
-	riscv_set_register(target, GDB_REGNO_TDATA2, trigger->address);
+MATCH_GE_LT:
+	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, true, &idx);
+	if (ret != ERROR_OK)
+		return ret;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_GE);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_ENABLED);
+	tdata2 = trigger->address;
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK) {
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
```

Comment:
Should there be some kind of warning that the requested trigger length isn't supported?
How often does this happen? If it happens for every watchpoint (when ge/lt aren't supported) it would only be annoying to print a message.

Reply:
A fallback is required on machines that do not support MATCH_LT or MATCH_GE. Fallback may result in some data modifications not triggering watchpoint.

Reply:
In my opinion, if this fallback gets hit, it would be good to:

* print a `LOG_WARNING()` message when this occurs for the first time for a given target
* print `LOG_DEBUG()` every time